Do not use unsafe restriding for subclasses

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88092
* __->__ #87610

This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Do not use unsafe restriding for subclasses

This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@bdhirsh is grabbing this.

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD reviewed (approved):
Nice

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -223,7 +223,13 @@ void scatter_meta_impl(
+      meta.set_output_raw_strided(0, self.sizes(), self.strides(), self.options());
+  } else {
+      meta.set_output_raw_strided(0, self.sizes(), {}, self.options());
+  }
```
@ezyang I'll wait for CI, but this fixes the errors locally. I wanted to ask what you think of this fix though. Here's the reason the test fails without this "fix":

<img width="1254" alt="image" src="https://user-images.githubusercontent.com/16311747/198374218-63f0b764-8780-4604-8bc0-f50738b439e4.png">

in the `functionalize(remove="mutations")` case where we keep views around, the "propagate mutations back to the base by running the inverse of every view" logic needs every intermediate `foo_updated` to have the same strides as `foo`.

That's not the case today though when we run `*_scatter` ops though - they always return contiguous tensors.

I think there are a few ways we could fix this:

(1) updated the scatter ops to do the thing described above. Technically to be completely correct, they should always return the same striding pattern as the input **even if** the input wasn't memory dense. Maybe I should just... do that? It felt like a pretty big change to `scatter` though.

(2) Let functionalization handle it, instead of mucking with the op directly - specifically, we could update `slice_inverse(..)` to look like:
```
slice_inverse(base, mutated_view, ...):
    tmp = torch.slice_scatter(base, mutated_view, ...)
    # we don't need to worry about proper striding when everything is guaranteed to be contiguous
    if (!remove_views):
        return tmp
    else:
        return tmp.as_strided(base.size(), base.stride(), base.storage_offset());
```

This feels better because we're not mucking with the contract of scatter ops directly, but downsides include:

(1) This won't actually properly handle the case where the input to the scatter op wasn't memory dense. Returning a non-memory-dense output would need to be done directly by the scatter op
(2) we'll end up tracing more `as_strided()` calls directly in the the graph, which feels bad (maybe not a big deal though)

bdhirsh replied:
FWIW, the failing test required me to patch `scatter.src`, because the test uses vmap, which happens to decompose `slice_scatter` into `scatter.src`. I would probably need to apply the same patch to the other four `*_scatter` ops though.

bdhirsh replied:
FWIW, I'm leaning towards (1) (beefing up this existing PR so that scatter properly propagates strides of non-densely packed inputs, and doing the same for the 4 other scatter ops)

ezyang replied:
My initial take is that the scatters should result in the input striding exactly. It is kind of inefficient if you have sparse views in the middle.

What is kind of annoying is that if you use as strided to compute the scatter, you never have to worry about discontiguous inputs in the middle, because you'll scatter straight to the dense input. So it is kind of an interesting question if compilers are better able to do the scatter if we decompose it or not.

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh committed:
Update on "Do not use unsafe restriding for subclasses"


This helps convert some accuracy errors into runtime errors,
which makes it easier to debug.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

mehtanirav commented:
@pytorchbot revert -reason [Internal breakages](https://www.internalfb.com/intern/sandcastle/job/36028797828925790/insights)

bdhirsh commented:
@pytorchbot rebase

mehtanirav commented:
@pytorchbot revert -m "[Internal breakages](https://www.internalfb.com/intern/sandcastle/job/36028797828925790/insights)" -c ghfirst

bdhirsh commented:
had trouble re-using the same PR to re-land, closing in favor of https://github.com/pytorch/pytorch/pull/88458
